tower/src/util/and_then.rs

```diff
@@ -0,0 +1,74 @@
+use std::task::{Context, Poll};
+
+use futures_core::TryFuture;
+use futures_util::{
+    future::{AndThen as AndThenFut, ErrInto as ErrIntoFut},
+    TryFutureExt,
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Service returned by the [`and_then`] combinator.
+///
+/// [`and_then`]: crate::util::ServiceExt::and_then
+#[derive(Clone, Debug)]
+pub struct AndThen<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> AndThen<S, F> {
+    /// Creates a new `AndThen` service.
+    pub fn new(inner: S, f: F) -> Self {
+        AndThen { f, inner }
+    }
+}
+
+impl<S, F, Request, Error, Fut> Service<Request> for AndThen<S, F>
+where
+    S: Service<Request>,
+    S::Error: Into<Error>,
+    F: FnOnce(S::Response) -> Fut + Clone,
+    Fut: TryFuture<Error = Error>,
+{
+    type Response = Fut::Ok;
+    type Error = Error;
+    type Future = AndThenFut<ErrIntoFut<S::Future, Error>, Fut, F>;
```

Comment:
let's change this to a named type re-exported from `util::future`, the way the other combinators do. I added a macro to make this easier --- if you rebase onto master, you can follow the example of PR #509

Reply:
This doesn't seem possible
```rust
opaque_future! {
    /// Response future from [`AndThen`] services.
    ///
    /// [`AndThen`]: crate::util::AndThen
    pub type AndThenFuture<F1, F2: TryFuture, N> = future::AndThen<future::ErrInto<F1, F2::Error>, F2, N>;
}
```

Is this acceptable?

```rust
opaque_future! {
    /// Response future from [`AndThen`] services.
    ///
    /// [`AndThen`]: crate::util::AndThen
    pub type AndThenFuture<F1, F2, F2Error, N> = future::AndThen<future::ErrInto<F1, F2Error>, F2, N>;
}
```

Reply:
Either is fine --- you're right that the macro doesn't support trait bounds currently. We could try to hack that in, but I think it would be fine to manually write our own newtype.

Remember that if you do that, you should write `pub struct AndThenFuture<...>(...)` and write a `Future` impl for it, rather than `pub type AndThenFuture...`. This is what the macro would expand to.

Reply:
How's that look?